datashader/core.py

```diff
@@ -227,65 +228,39 @@ def raster(self,
 
         resample_methods = dict(nearest=0, bilinear=1)
 
         if resample_method not in resample_methods.keys():
             raise ValueError('Invalid resample method: options include {}'.format(list(resample_methods.keys())))
 
-        # setup output array
-        full_data = np.empty(shape=(self.plot_width, self.plot_height)).astype(source.profile['dtype'])
-        full_xs = np.linspace(self.x_range[0], self.x_range[1], self.plot_width)
-        full_ys = np.linspace(self.y_range[0], self.y_range[1], self.plot_height)
-        attrs = dict(res=source.res[0], nodata=source.nodata)
-        full_arr = DataArray(full_data,
-                             coords=[('x', full_xs), ('y', full_ys)],
-                             attrs=attrs)
-
-        # handle out-of-bounds case
-        if (self.x_range[0] >= source.bounds.right or
-            self.x_range[1] <= source.bounds.left or
-            self.y_range[0] >= source.bounds.top or
-            self.y_range[1] <= source.bounds.bottom):
-            return full_arr
+        res = calc_res(source)
+        left, bottom, right, top = calc_bbox(source.x.values, source.y.values, res)
 
         # window coodinates
-        xmin = max(self.x_range[0], source.bounds.left)
-        ymin = max(self.y_range[0], source.bounds.bottom)
-        xmax = min(self.x_range[1], source.bounds.right)
-        ymax = min(self.y_range[1], source.bounds.top)
+        xmin = max(self.x_range[0], left)
+        ymin = max(self.y_range[0], bottom)
+        xmax = min(self.x_range[1], right)
+        ymax = min(self.y_range[1], top)
 
         width_ratio = (xmax - xmin) / (self.x_range[1] - self.x_range[0])
         height_ratio = (ymax - ymin) / (self.y_range[1] - self.y_range[0])
 
+        if np.isclose(width_ratio, 0) or np.isclose(height_ratio, 0):
+            raise ValueError('Canvas x_range or y_range values do not match closely-enough with the data source to be able to accurately rasterize. Please provide ranges that are more accurate.')
+
         w = int(np.ceil(self.plot_width * width_ratio))
         h = int(np.ceil(self.plot_height * height_ratio))
 
-        rmin, cmin = source.index(xmin, ymin)
-        rmax, cmax = source.index(xmax, ymax)
-
-        if use_overviews:
-            data = np.empty(shape=(h, w)).astype(source.profile['dtype'])
-            data = source.read(band, out=data, window=((rmax, rmin), (cmin, cmax)))
-        else:
-            data = source.read(band, window=((rmax, rmin), (cmin, cmax)))
-
-        is_int = np.issubdtype(data.dtype, np.integer)
-        data[data == np.array(source.nodata)] = 0 if is_int else np.nan
-
-        # TODO: this resize should go away once rasterio has overview resample
-        data = resize(data,
-                      (h, w),
-                      order=resample_methods[resample_method],
-                      preserve_range=True)
+        data = resample_2d(source.values[0].astype(np.float32), w, h)
```

Comment:
Both ds and us methods seem important, and better to expose them now than have to change later.